### PR TITLE
fix: EXPOSED-407 compositeMoney() nullability definition issues

### PR DESCRIPTION
**Pull request by bog-walk**
Defining a composite money column as nullable returns a different type:
```kt
// non-nullable column
val money: CompositeMoneyColumn<BigDecimal, CurrencyUnit, MonetaryAmount> = compositeMoney()

// actual nullable column
val nullableMoney: CompositeMoneyColumn<MonetaryAmount?> = compositeMoney().nullable()

// expected nullable column
val money: CompositeMoneyColumn<BigDecimal?, CurrencyUnit?, MonetaryAmount?> = compositeMoney().nullable()
```

The current setup works as long as the insert value is a composite `Money` value, as the individual columns will not be able to be resolved.

If the component values (`BigDecimal` and `CurrencyUnit`) are inserted individually into the component columns, it will only work for either a non-nullable column or a nullable column with a non-null value. This happens because each component column is correctly nullable, but not the actual composite expression itself, resulting in the error:
```
Can't set null value to non-nullable CompositeMoneyColumn column
```

This fix adds an overload so that `compositeMoney().nullable()` returns the expected type and adds the ability to modify the internal `nullable` property in `BiCompositeColumns`.

None of the current tests included inserts/updates that targeted the individual component columns, so this has been included too.

### CODE REVIEW

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/CompositeColumn.kt [hunk 1]
```diff
@@ -39,8 +39,12 @@ abstract class BiCompositeColumn<C1, C2, T>(
     /** Transformation that receives the column's composite value and returns the parsed values of the underlying columns. */
     val transformFromValue: (T) -> Pair<C1?, C2?>,
     /** Transformation that receives the retrieved values of [column1] and [column2] and returns a composite value. */
-    val transformToValue: (Any?, Any?) -> T
+    val transformToValue: (Any?, Any?) -> T,
+    nullable: Boolean = false
 ) : CompositeColumn<T>() {
+    init {
+        this.nullable = nullable
+    }
```
Comment (bog-walk) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/CompositeColumn.kt:47]: The alternative could be to make the `nullable` property `protected` in `CompositeColumn` above and perform this set in `CompositeMoneyColumn` directly. I thought it might be good for users that extend this class to have the option to alter nullability.

File: exposed-money/src/main/kotlin/org/jetbrains/exposed/sql/money/CompositeMoneyColumn.kt [hunk 2]
```diff
@@ -36,12 +40,26 @@ class CompositeMoneyColumn<T1 : BigDecimal?, T2 : CurrencyUnit?, R : MonetaryAmo
...
         currency = Column(table, currencyName, CurrencyColumnType())
     )
+
+/** Marks this [CompositeMoneyColumn] as nullable. */
+@Suppress("UNCHECKED_CAST")
+fun <T1 : BigDecimal, T2 : CurrencyUnit, R : MonetaryAmount> CompositeMoneyColumn<T1, T2, R>.nullable(): CompositeMoneyColumn<T1?, T2?, R?> {
+    return with(amount.table) {
+        (this@nullable as BiCompositeColumn<T1, T2, R>).nullable() as CompositeMoneyColumn<T1?, T2?, R?>
+    }
+}
```
Comment (bog-walk) [inline exposed-money/src/main/kotlin/org/jetbrains/exposed/sql/money/CompositeMoneyColumn.kt:65]: `.nullable()` must be called within the context of a `Table`.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/CompositeColumn.kt (hunk 1 above)
Comment (e5l) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/CompositeColumn.kt:47]: I would wait for request to do this

File: exposed-money/src/test/kotlin/org/jetbrains/exposed/sql/money/MoneyTests.kt [hunk 3]
```diff
@@ -78,27 +88,61 @@ open class MoneyBaseTest : DatabaseTestsBase() {
...
+                .where { tester.money.amount.isNotNull() and tester.nullableMoney.amount.isNotNull() }
+                .single()
+            assertEquals(currencyUnit, result2[tester.money.currency])
+            assertEquals(currencyUnit, result2[tester.nullableMoney.currency])
+
+            // manual composite columns should still accept composite values
+            val compositeMoney = Money.of(BigDecimal(10), "CAD")
+            tester.insert {
+                it[money] = compositeMoney
+                it[nullableMoney] = null
```
Comment (e5l) [inline exposed-money/src/test/kotlin/org/jetbrains/exposed/sql/money/MoneyTests.kt:135]: I would also add tests when we skip setting null explicitly

**Review by e5l — APPROVED** [review]
Nice catch! LGTM

File: exposed-money/src/test/kotlin/org/jetbrains/exposed/sql/money/MoneyTests.kt [hunk 4]
```diff
@@ -107,11 +159,27 @@ open class MoneyBaseTest : DatabaseTestsBase() {
             assertEquals(toInsert, inserted)
         }
     }
+
+    private fun testInsertedAndSelectByComponentColumns(toInsert: Money?) {
+        withTables(Account) {
```
Comment (obabichevjb) [inline exposed-money/src/test/kotlin/org/jetbrains/exposed/sql/money/MoneyTests.kt:164]: It's minor, but I'm usually trying to call `withTables`/`withDb` on the root test function, mostly because of `excludeSettings` to make it clear, which DBs are involved.

And usually, if I do `assert()` inside function, I call this function also `assertSomething()` to show that it may stop the test

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/CompositeColumn.kt [hunk 5]
```diff
@@ -39,8 +39,12 @@ abstract class BiCompositeColumn<C1, C2, T>(
     /** Transformation that receives the column's composite value and returns the parsed values of the underlying columns. */
     val transformFromValue: (T) -> Pair<C1?, C2?>,
     /** Transformation that receives the retrieved values of [column1] and [column2] and returns a composite value. */
-    val transformToValue: (Any?, Any?) -> T
+    val transformToValue: (Any?, Any?) -> T,
```
Comment (obabichevjb) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/CompositeColumn.kt:42]: Is it intentional to define `transformToValue` as `(Any?, Any?) -> T`. Does it mean that the user should manually call `columnType.valueFromDB()` inside that function?

I tried to replace it with `val transformToValue: (C1?, C2?) -> T?` and update `restoreValueFromParts()` to 
```
override fun restoreValueFromParts(parts: Map<Column<*>, Any?>): T? {
  val v1 = parts[column1]?.let { column1.columnType.valueFromDB(it) }
  val v2 = parts[column2]?.let { column2.columnType.valueFromDB(it) }
  val result = transformToValue(v1, v2)
  check(result != null || nullable) {
      "Null value received from DB for non-nullable ${this::class.simpleName} column"
  }
  return result
}
```

and it seems to be working, but I don't know which use cases I loast with such update. But after fixing `CompositeMoneyColumn` I also got green tests (without modifying tests itself).

Anyway the current variant is also ok to me, but `(Any, Any)` surrounded by generics looks alien)

**Review by obabichevjb — APPROVED** [review]

File: exposed-money/src/test/kotlin/org/jetbrains/exposed/sql/money/MoneyTests.kt (hunk 4 above)
Comment (bog-walk) [reply exposed-money/src/test/kotlin/org/jetbrains/exposed/sql/money/MoneyTests.kt:164]: I agree that it is nicer when tests easily make it clear from the start what tables are being used, as well as what db are being excluded. I'll change all the old code.
